test/python/primitives/test_backend_estimator_dev.py

```diff
@@ -0,0 +1,924 @@
+    AbelianDecomposer,
+    BackendEstimator,
+    NaiveDecomposer,
+)
+from qiskit.providers import Backend, Options, JobV1
+from qiskit.providers.fake_provider import FakeNairobi, FakeNairobiV2
+from qiskit.quantum_info.operators import SparsePauliOp
+from qiskit.quantum_info.operators.symplectic.pauli import Pauli
+from qiskit.quantum_info.operators.symplectic.pauli_list import PauliList
+from qiskit.result import Counts, Result
+from qiskit.transpiler import Layout, PassManager
+from qiskit.transpiler.passes import ApplyLayout, SetLayout
+from qiskit.test import QiskitTestCase
+
+BACKENDS = [FakeNairobi(), FakeNairobiV2()]
+
+
+################################################################################
+## AUXILIARY
+################################################################################
+def measurement_circuit_examples() -> Iterator[tuple[list[str], QuantumCircuit]]:
+    """Generator of commuting Paulis and corresponding measurement circuits.
+
+    Yields:
+        - List of commuting Pauli strings
+        - Quantum circuit to measure said Paulis
+    """
+    I = QuantumCircuit(1, 1)  # pylint: disable=invalid-name
+    I.measure(0, 0)
+    yield ["I", "Z"], I
+
+    X = QuantumCircuit(1, 1)  # pylint: disable=invalid-name
+    X.h(0)
+    X.measure(0, 0)
+    yield ["X", "I"], X
+
+    Y = QuantumCircuit(1, 1)  # pylint: disable=invalid-name
+    Y.sdg(0)
+    Y.h(0)
+    Y.measure(0, 0)
+    yield ["Y", "I"], Y
+
+    Z = QuantumCircuit(1, 1)  # pylint: disable=invalid-name
+    Z.measure(0, 0)
+    yield ["Z", "I"], Z
+
+    II = QuantumCircuit(2, 1)  # pylint: disable=invalid-name
+    II.measure(0, 0)
+    yield ["II"], II
+
+    IY = QuantumCircuit(2, 1)  # pylint: disable=invalid-name
+    IY.sdg(0)
+    IY.h(0)
+    IY.measure(0, 0)
+    yield ["IY", "II"], IY
+
+    XY = QuantumCircuit(2, 2)  # pylint: disable=invalid-name
+    XY.h(1)
+    XY.sdg(0)
+    XY.h(0)
+    XY.measure([0, 1], [0, 1])
+    yield ["XY", "II", "XI", "IY"], XY
+
+    XX = QuantumCircuit(2, 2)  # pylint: disable=invalid-name
+    XX.h(1)
+    XX.h(0)
+    XX.measure([0, 1], [0, 1])
+    yield ["XX", "IX", "XI", "II"], XX
+
+    ZZ = QuantumCircuit(2, 2)  # pylint: disable=invalid-name
+    ZZ.measure([0, 1], [0, 1])
+    yield ["ZZ", "IZ", "ZI", "II"], ZZ
+
+    XYZ = QuantumCircuit(3, 3)  # pylint: disable=invalid-name
+    XYZ.h(2)
+    XYZ.sdg(1)
+    XYZ.h(1)
+    XYZ.measure([0, 1, 2], [0, 1, 2])
+    yield ["XYZ", "XII", "IYI", "IIZ", "XIZ", "III"], XYZ
+
+    YIX = QuantumCircuit(3, 2)  # pylint: disable=invalid-name
+    YIX.sdg(2)
+    YIX.h(2)
+    YIX.h(0)
+    YIX.measure([0, 2], [0, 1])
+    yield ["YIX", "IIX", "YII", "III"], YIX
+
+    IXII = QuantumCircuit(4, 1)  # pylint: disable=invalid-name
+    IXII.h(2)
+    IXII.measure(2, 0)
+    yield ["IXII", "IIII"], IXII
+
+
+def circuit_composition_examples() -> Iterator[
+    tuple[QuantumCircuit, QuantumCircuit, QuantumCircuit]
+]:
```

Comment:
This is a one-use function for parametrisation, and the objects it creates are very easy to build - it would be better for it just to be inline in the `data` parametrisation for the specific test case.  That would keep the data local to the test case.

Personally, I also find the building here very hard to read and track.  I have to chase the logic through 5 different functions, which result in three different `QuantumCircuit`s being created for the test case, and I have to remember the details of all of them to know whether I believe the test is correct or not.  In general, test code should be _obviously_ correct just from reading the test function, and I definitely wouldn't know that here.

(There's also the thing that all this complexity seems to just be in service of testing a private method anyway, which I already argued against.)

Reply:
Somewhat answered here: https://github.com/Qiskit/qiskit-terra/pull/9074#discussion_r1028436793